_posts/2021-10-08-grokking-clojure-transducers.md

```diff
@@ -0,0 +1,400 @@
+
+Here's an example of using `conj`:
+
+```clojure
+(conj [1 2] 3)
+;;          ^ input
+;;    ^^^^^ accumulated result
+;;=> [1 2 3]
+;;   ^^^^^^^ new result
+```
+
+Since `conj` is a reducing function, let's try using it with `reduce`. Here, we use `conj` to add every number in the given Clojure vector (`[1 2 2 3 1]`) into an empty hash set (`#{}`):
+
+```clojure
+(reduce conj #{} [1 2 2 3 1])
+;;=> #{1 3 2}
+```
+
+Cool! We got what we expected: a set that contains each distinct number in the input collection.
+
+Now that we know what reducing functions are, let's discuss how they relate to transducers.
+
+<!-- ## Making our first transducer -->
+
+In essence, a **transducer** is *a function that takes a reducing function* (like `conj`) and *turns it into a new, more awesome reducing function*. Let's give it a try. Let's make a transducer called `inc-transducer`:
+
+```clojure
+(defn inc-transducer
+  "Given a reducing function rf, return a new reducing function that increments
+  every input it receives, then calls rf with the result and the incremented
+  input."
+  ;; rf stands for "reducing function"
+  [rf]
+  ;; this here's a new reducing function
+  (fn [result input]
+    ;; here we call the original reducing function
+    (rf result (inc input))))
+
+;; If you already know transducers, you'll notice that inc-transducer doesn't
+;; actually have everything a real, production-grade transducer needs to have.
+;; Don't worry, we'll get there.
+```
+
+We've made our first transducer! `inc-transducer` takes a reducing function like `conj` and returns a modified version of it. Let's give it a try:
+
+```clojure
+(def inc-then-conj (inc-transducer conj))
+;;=> #'user/inc-then-conj
+(inc-then-conj [1 2] 3)
+;;=> [1 2 4]
+```
+
+What is `inc-then-conj` supposed to do again? Let's recap. It should:
+
+1. Increment the input `3` by one to get `4`.
+2. Add the incremented input `4` into the result to get `[1 2 4]`.
+
+Looks like it works! Can we use it with `reduce`, like we did with `conj`?
+
+```clojure
+(reduce inc-then-conj [] [1 2 3 4 5])
+;;=> [2 3 4 5 6]
+```
+
+Boom! When used with `reduce`, `inc-then-conj` incremented every input, then added each incremented input into the result vector (`[]`).
+
+Of course, we can give `inc-transducer` *any* reducing function, not just `conj`. Let's try to use it to transform some other reducing function. Let's pick `+`, which, happily, is also a reducing function:
+
+```clojure
+(reduce (inc-transducer +) 0 [1 2 3 4 5])
+;;=> 20
+```
+
+It worked! When we used `inc-transducer` with `+`, for each number in the input collection, it incremented the number, then added it into the accumulated sum.
+
+However... what if there comes a day when we want to do something else than increment a number? That's all `inc-transducer` lets us do, increment numbers! What do?
```

Comment:
What **to** do?

Reply:
Believe it or not, this was actually intentional. I totally thought I was using this cool meme, but it appears I completely made it up. Will fix. 👍🏻